cardano-api-test/test/Test/Cardano/Api/TxBody.hs

```diff
@@ -0,0 +1,263 @@
+    }
+
+normalizeInsCollateral :: TxInsCollateral era -> TxInsCollateral era
+normalizeInsCollateral = \case
+  TxInsCollateralNone -> TxInsCollateralNone
+  -- for original
+  TxInsCollateral _ ins | null ins -> TxInsCollateralNone
+  -- for roundtrip
+  TxInsCollateral support ins -> TxInsCollateral support $ sort ins
+
+normalizeMetadata :: TxMetadataInEra era -> TxMetadataInEra era
+normalizeMetadata = \case
+  TxMetadataInEra _ (TxMetadata m) | null m -> TxMetadataNone
+  other                                     -> other
+
+normalizeAuxScripts :: TxAuxScripts era -> TxAuxScripts era
+normalizeAuxScripts = \case
+  TxAuxScripts _       []      -> TxAuxScriptsNone
+  TxAuxScripts support scripts ->
+    -- sorting uses script versions, hence sort after upgrade
+    TxAuxScripts support $ sort $ map upgradeScriptInEra scripts
+  other -> other
+
+normalizeWithdrawals :: TxWithdrawals ViewTx era -> TxWithdrawals ViewTx era
+normalizeWithdrawals = \case
+  TxWithdrawals _ [] -> TxWithdrawalsNone
+  other              -> other
+
+normalizeCertificates :: TxCertificates ViewTx era -> TxCertificates ViewTx era
+normalizeCertificates = \case
+  TxCertificates _ [] _ -> TxCertificatesNone
+  other                 -> other
+
+normalizeMintValue :: TxMintValue ViewTx era -> TxMintValue ViewTx era
+normalizeMintValue = \case
+  TxMintValue _ v _ | v == mempty -> TxMintNone
+  other                           -> other
+
+normalizeUpdateProposal :: TxUpdateProposal era -> TxUpdateProposal era
+normalizeUpdateProposal = \case
+  TxUpdateProposalNone -> TxUpdateProposalNone
+  TxUpdateProposal support (UpdateProposal update epoch) ->
+    TxUpdateProposal
+      support
+      (UpdateProposal (normalizeUpdateParameters <$> update) epoch)
+
+normalizeUpdateParameters
+  :: ProtocolParametersUpdate -> ProtocolParametersUpdate
+normalizeUpdateParameters params =
+  params
+    { protocolUpdateCollateralPercent =
+        zeroIsNothing $ protocolUpdateCollateralPercent params
+    , protocolUpdateMaxBlockExUnits =
+        zeroExecutionIsNothing $ protocolUpdateMaxBlockExUnits params
+    , protocolUpdateMaxCollateralInputs =
+        zeroIsNothing $ protocolUpdateMaxCollateralInputs params
+    , protocolUpdateMaxTxExUnits =
+        zeroExecutionIsNothing $ protocolUpdateMaxTxExUnits params
+    , protocolUpdateMaxValueSize =
+        zeroIsNothing $ protocolUpdateMaxValueSize params
+    , protocolUpdatePrices = zeroPricesIsNothing $ protocolUpdatePrices params
+    , protocolUpdateUTxOCostPerWord =
+        zeroIsNothing $ protocolUpdateUTxOCostPerWord params
+    }
+  where
+    zeroIsNothing :: (Eq a, Num a) => Maybe a -> Maybe a
+    zeroIsNothing = \case
+      Just 0 -> Nothing
+      x      -> x
+    zeroPricesIsNothing = \case
+      Just (ExecutionUnitPrices 0 0) -> Nothing
+      x                              -> x
+    zeroExecutionIsNothing = \case
+      Just (ExecutionUnits 0 0) -> Nothing
+      x                         -> x
```

Comment:
These seem very suspicious. There should be a difference between setting a param to zero and not changing it at all. This seems wrong, or working around a problem elsewhere. What was going wrong here? Perhaps it's a real bug.